tests/include/mock_custom_nrf_modem_at.h

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#ifndef MOCK_CUSTOM_NRF_MODEM_AT_H_
+#define MOCK_CUSTOM_NRF_MODEM_AT_H_
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @file mock_custom_nrf_modem_at.h
+ * @brief Public APIs for the Custom nrf_modem_at mock.
+ * @defgroup mock_custom_nrf_modem_at Custom mock for nrf_modem_at
+ * @{
+ */
+
+/**
+ * @brief Initialize mocking variables for nrf_modem_at_scanf() calls.
+ *
+ * This is planned to be used in setUp() function.
+ */
+void mock_custom_nrf_modem_at_Init(void);
+/**
+ * @brief Verify that nrf_modem_at_scanf() calls where executed as expected.
+ *
+ * This is planned to be used in tearDown() function.
+ */
+void mock_custom_nrf_modem_at_Verify(void);
+
+/**
+ * @brief Mock for nrf_modem_at_scanf().
+ */
+int __wrap_nrf_modem_at_scanf(const char *cmd, const char *fmt, ...);
+
+/**
+ * @brief Set expected parameter and return values for non-variable arguments.
+ */
+void __wrap_custom_nrf_modem_at_scanf_ExpectAndReturn(
+	const char *cmd, const char *fmt, int return_value);
+
+/**
+ * @brief Indicate that previous __wrap_custom_nrf_modem_at_scanf_*() calls are complete for
+ * one nrf_modem_at_scanf() call.
+ */
+void __wrap_custom_nrf_modem_at_scanf_Ready(void);
```

Comment:
I wonder if this could be a side-effect of calling ExpectAndReturn, e.g, call ReturnVarg N times and then call ExpectAndReturn.

Reply:
I wanted to have ExpectAndReturn first and then ReturnVarg calls because that is more aligned with Cmock's style of having ExpectAndReturn and IgnoreArg calls. But I guess it would be possible to come up with something in ExpectAndReturn so that we wouldn't need Ready call at all. I need to try this when I have time.

Reply:
Yeah, if you think keeping the call order will reduce the chance of confusion or miss-use, that makes sense to me

In that case, perhaps it could be named _ReturnVargEnd(), and we could make it only required if there has been a ReturnVarg() call since the last ExpectAndReturn

I am fine with it as it is, though, these are just ideas

Reply:
I think I was able to change ExpectAndReturn so that I got rid of Ready function without removing any of the checks about too many or too few function calls.

Reply:
I may have approved too soon, whoops, I hadn't noticed you added these changes

I am concerned, there doesn't seem to be a way for the arguments for two consecutive calls to nrf_modem_at_scanf to be distinguished. I am worried that this might create false positives or confusing negatives.

For example, if someone is executing a test that calls a 3-param call then a 4-param call, and those get called in the wrong order.

Maybe it would be a good idea to add an argument to ExpectAndReturn that specifies the number of var arguments to expect to be passed.

Reply:
You can set `CONFIG_MOCK_CUSTOM_NRF_MODEM_AT_SCANF_CALL_COUNT` number of different `nrf_modem_at_scanf()` calls to be expected and each of those calls can have `CONFIG_MOCK_CUSTOM_NRF_MODEM_AT_SCANF_VARGS_COUNT` number of arguments.
Removal of `__wrap_custom_nrf_modem_at_scanf_Ready()` didn't change any functionality or affect on this behavior.

As you probably know, the way it works is that `__wrap_custom_nrf_modem_at_scanf_ExpectAndReturn()` starts a new expected `nrf_modem_at_scanf()` call and all `__wrap_custom_nrf_modem_at_scanf_ReturnVarg_*()` calls after that and before next `__wrap_custom_nrf_modem_at_scanf_ExpectAndReturn()` call will be assigned to the specific `nrf_modem_at_scanf()` call.

An example of two expected `nrf_modem_at_scanf()` calls can be seen in `test_lte_lc_conn_eval_params_coneval_*()`.

> Maybe it would be a good idea to add an argument to ExpectAndReturn that specifies the number of var arguments to expect to be passed.

I don't know how I could verify the number of arguments in `__wrap_nrf_modem_at_scanf()` because `va_arg()` and other functions handling variadic argument list don't provide means to check the number of arguments. `va_arg()` returns undefined return value if `va_list` doesn't have more arguments.
So at the moment, if you give too few expected return vargs, the rest are just ignored and I've used this in some failure cases when I'm not interested on the arguments. If you give too many expected arguments, behavior is undefined and when I've tried this, I get segmentation fault. In theory, this could give false positive if va_arg() would return zero etc. and that would be expected by the code calling `nrf_modem_at_scanf()` and hence the test would still pass.

> For example, if someone is executing a test that calls a 3-param call then a 4-param call, and those get called in the wrong order.

In this case, I would hope that the expected command or response format would be different and those assertions would fail the test. If your code calls sequentially `nrf_modem_at_scanf()` with same command and format with 3 parameters and immediately with 4 parameters, the mock may give you false positive. Given the needed preconditions for this to happen, I'd say that you shouldn't be able to get to this issue.

Reply:
Forgive me, I was deeply mistaken about variadics in C. Apparently I was conflating with some other language. I thought there was a way to get the total count of supplied args. In fact I badly misread your code based on this assumption. My fault for not reading `__wrap_nrf_modem_at_scanf` in detail.

Since there is no way to count var args in C, my point is completely moot. You've done the best that can be done.


> As you probably know, the way it works is that __wrap_custom_nrf_modem_at_scanf_ExpectAndReturn() starts a new expected nrf_modem_at_scanf() call and all __wrap_custom_nrf_modem_at_scanf_ReturnVarg_*() calls after that and before next __wrap_custom_nrf_modem_at_scanf_ExpectAndReturn() call will be assigned to the specific nrf_modem_at_scanf() call.

I think there should be a statement on this in the header docs for the function. Something like:

> Expect modem_at_scanf to be called and set the expected parameter and return values for non-variable arguments. Must be called before ReturnVarg calls associated with the same modem_at_scanf call.

At any rate, I'll approve, sorry for the foolish hold-up.

Reply:
I've added some text to the function description. When you are ok with the resolution, please close this conversation.

Reply:
I love what you wrote, LGTM!